### PR TITLE
Prevent modifications to non-managed Hive tables

### PR DESCRIPTION
**Pull request by electrum**

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 1]
```diff
@@ -1187,11 +1187,11 @@ public OptionalLong metadataDelete(ConnectorSession session, ConnectorTableHandl
             throw new TableNotFoundException(handle.getSchemaTableName());
         }
 
+        if (!table.get().getTableType().equals(TableType.MANAGED_TABLE.toString())) {
```
Comment (cberner) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: static import `MANAGED_TABLE`?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveWriteUtils.java [hunk 2]
```diff
@@ -293,6 +294,10 @@ public static Object getField(Type type, Block block, int position)
 
     public static void checkTableIsWritable(Table table)
     {
+        if (!table.getTableType().equals(TableType.MANAGED_TABLE.toString())) {
```
Comment (cberner) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveWriteUtils.java]: static import `MANAGED_TABLE`?

**Comment by cberner** [comment]
Looks good

**Comment by KBP-TDC** [comment]
@electrum : is there any impact on Hive tables stored in S3? Will your change introduce any regression for current Presto users in AWS?
CC: @nezihyigitbasi

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java (hunk 1 above)
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: I avoided that to skip changing unrelated lines. Also, since this a string comparison, having the full name doesn't seem unreasonable

**Comment by electrum** [comment]
@KBP-TDC It might cause a regression, but allowing it in the first place was incorrect. A external table should be a read-only view of externally managed data. Otherwise, the behavior doesn't make sense: creating a new partition writes data, but dropping the partition only drops the metadata, not the data. Thus, the following operation fails (in a very confusing way), which is what prompted this change:
```sql
INSERT INTO ... 
DELETE FROM ...
INSERT INTO ...

**Comment by KBP-TDC** [comment]
@electrum is Presto's current behavior different than Hive's? In any case, I think we should have a property switch to enable the previous behavior for folks who already adopted Presto and rely on current semantics.
CC: @martint @mattsfuller

**Comment by electrum** [comment]
Hive has even weirder behavior. @haozhun can explain

**Comment by haozhun** [comment]
In Hive,
* For “insert into”
  * If partition doesn’t exist (in metastore), but the partition directory exists, old data in the directory won’t be removed. This means you are going to have data in the partition that were never inserted.
    * Presto refuses to insert in such case.
  * If partition exists (in metastore), Hive's behavior is sort of reasonable. But then, the data that got created on the disk will not be deleted when the partition later get dropped in Hive.
    * Presto does the same thing right now. This pull request is trying to change this.
* For “insert overwrite”
  * If partition doesn’t exist (in metastore), but the partition directory exists, old data in the directory won’t be removed. This means you are going to have data in the partition that were never inserted, and are supposedly overwritten.
    * In Presto, this needs to be simulated with "delete" followed by "insert". Presto will fail on insert (explained above).
  * If partition exists (in metastore), Hive's behavior is somewhat reasonable. But then, Hive actually deletes data from storage for external table, which is not supposed to happen.
    * In Presto, this needs to be simulated with "delete" followed by "insert". Presto will fail on insert (explained above).

**Comment by KBP-TDC** [comment]
@haozhun Thanks, I do see how all of this is confusing. Effectively, this PR is trying to make EXTERNAL tables read-only which is much safer behavior. I do worry about inconsistency with Hive b/c user may be confused. Hence, my request to have this configurable.

I might be missing something but aren't all S3-backed tables external by nature? If we change the current Presto behavior without ability to fallback to the old way of dealing with INSERTs, we may impact current Presto users in production. Not sure whether it is easy to "upgrade" an existing external table into a managed one in Hive Metastore. Let's get input from folks running in AWS if things aren't clear.

**Comment by haozhun** [comment]
Yes. Let's see what @nezihyigitbasi and @zhenxiao thinks.

**Comment by zhenxiao** [comment]
@haozhun @KBP-TDC @electrum @cberner I think it is safe to prevent modifications to external tables on S3. @nezihyigitbasi Do we have pipelines using Presto's insert/delete?

**Comment by electrum** [comment]
There's nothing special about S3 tables. They simply have a location URL that starts with s3:// instead of HDFS://. External is an orthogonal issue.

**Comment by KBP-TDC** [comment]
@electrum I have checked with a few users we support and this PR is not going to affect us.
Given that tests on Travis failed, I presume you want to update current test cases and possibly add some new ones.

**Comment by nezihyigitbasi** [comment]
@KBP-TDC you can have managed s3 tables too. With external tables when the table is dropped only the metadata is removed from hive metastore and the data remains on s3, for managed tables the data on s3 also gets removed. We can also change an external table to a managed by altering the table and setting some table properties on it.

@zhenxiao currently we don't AFAIK, but we may have in the future.

I understand the confusion of insert -> delete -> insert failing with external tables and this change makes sense to prevent this confusion. I think overall this change is good as the behavior shouldn't be like this in the first place. And if someone wants Presto to write/delete to/from their tables they need to create managed tables (or convert external tables to managed) and this makes sense to me too.

**Comment by electrum** [comment]
@KBP-TDC The tests definitely need to be updated. I was surprised we have tests that insert into external tables.

**Comment by KBP-TDC** [comment]
@nezihyigitbasi @zhenxiao Thanks for the clarification. This PR looks like safe to go ahead with then.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java (hunk 1 above)
Comment (kokosing) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: Please add a test that queries fails with proper message when a `DELETE` or `INSERT` is issued on external tables. It could be either unit or product test.

**Comment by kokosing** [comment]
Superseded by: #5443

**Comment by BigDataDaddy** [comment]
I'm troubled to learn that this change will effectively break designs like one we have where multiple components, Spark, Hive, and Presto, all cooperate to manage an external tables files and partitions in a file system, like HDFS or EMRFS (i.e. S3) in a coordinated way. I agree that default settings should err on a commonly understood expected behavior, but configuration properties to relax newly restricted behavior or emulate original behaviors that existing system designs depend on would prevent regressions and breaking existing systems.

**Comment by electrum** [comment]
@BigDataDaddy This change shouldn't break that design. The external flag is a property of the table in the metastore. It says whether or not the metastore "owns" the data on disk. You can have Spark, Hive and Presto all working against the same managed table (within the constraints of the Hive concurrency model).

**Comment by jcap** [comment]
@electrum curious about the bug comment. I don't see much in the way of Hive indicating this is a problem. We were surprised to see this addition to the change logs.

**Comment by electrum** [comment]
It's a bug because allowing data modification violates the conceptual model (and the intent) of external tables.
